rust/psibase_macros/src/service_macro.rs

```diff
@@ -491,6 +544,345 @@ fn process_mod(
     .into()
 } // process_mod
 
+struct PkIdentData {
+    ident: Ident,
+    ty: proc_macro2::TokenStream,
+    call_ident: proc_macro2::TokenStream,
+}
+
+impl PkIdentData {
+    fn new(
+        ident: Ident,
+        ty: proc_macro2::TokenStream,
+        call_ident: proc_macro2::TokenStream,
+    ) -> Self {
+        Self {
+            ident,
+            ty,
+            call_ident,
+        }
+    }
+}
+
+struct SkIdentData {
+    ident: Ident,
+    idx: u8,
+    ty: proc_macro2::TokenStream,
+}
+
+impl SkIdentData {
+    fn new(ident: Ident, idx: u8, ty: proc_macro2::TokenStream) -> Self {
+        Self { ident, idx, ty }
+    }
+}
+
+fn process_service_tables(
+    psibase_mod: &proc_macro2::TokenStream,
+    table_record_struct_name: &Ident,
+    items: &mut Vec<Item>,
+    table_idxs: &Vec<usize>,
+) -> u16 {
+    let mut pk_data: Option<PkIdentData> = None;
+    let mut secondary_keys = Vec::new();
+    let mut table_options: Option<TableOptions> = None;
+
+    for idx in table_idxs {
+        match &mut items[*idx] {
+            Item::Struct(s) => {
+                process_table_attrs(s, &mut table_options, psibase_mod);
+                process_table_fields(s, &mut pk_data);
+            }
+            Item::Impl(i) => process_table_impls(i, &mut pk_data, &mut secondary_keys),
+            item => abort!(item, "Unknown table item to be processed"),
+        }
+    }
+
+    if pk_data.is_none() {
+        abort!(
+            table_record_struct_name,
+            "Table record has not defined a primary key"
+        );
+    }
+
+    let pk_data = pk_data.unwrap();
+    let pk_ty = pk_data.ty;
+    let pk_call_ident = pk_data.call_ident;
+
+    secondary_keys.sort_by_key(|sk| sk.idx);
+    let mut sks_fns = quote! {};
+    let mut sks = quote! {};
+    let mut last_idx = 0;
+    for secondary_key in secondary_keys.iter() {
+        let sk_ident = &secondary_key.ident;
+        let sk_idx = secondary_key.idx;
+        let sk_ty = &secondary_key.ty;
+
+        // check duplicates, since it's sorted we can check the last idx
+        if sk_idx as usize > secondary_keys.len() {
+            abort!(sk_ident, format!("Secondary index number is greater than the quantity of defined secondary indexes"));
```

Comment:
"Missing secondary index {}; indexes may not have gaps and may not be removed or reordered"